src/lib/storage/index/partial_hash/flat_map_iterator.cpp

```diff
@@ -0,0 +1,54 @@
+#include "flat_map_iterator.hpp"
+
+#include "flat_map_iterator_impl.hpp"
+
+namespace hyrise {
+
+FlatMapIterator::FlatMapIterator(std::unique_ptr<BaseFlatMapIteratorImpl>&& index_iterator)
+    : _impl(std::move(index_iterator)) {}
+
+FlatMapIterator::FlatMapIterator(const FlatMapIterator& other) : _impl(other._impl->clone()) {}
+
+FlatMapIterator& FlatMapIterator::operator=(const FlatMapIterator& other) {
+  if (&other != this) {
+    _impl = other._impl->clone();
+  }
+
+  return *this;
+}
+
+FlatMapIterator::reference FlatMapIterator::operator*() const {
+  return _impl->operator*();
+}
+
+FlatMapIterator& FlatMapIterator::operator++() {
+  _impl->operator++();
+  return *this;
+}
+
+bool FlatMapIterator::operator==(const FlatMapIterator& other) const {
+  return _impl->operator==(*other._impl);
+}
+
+bool FlatMapIterator::operator!=(const FlatMapIterator& other) const {
+  return _impl->operator!=(*other._impl);
+}
+
+template <typename DataType>
+FlatMapIterator FlatMapIterator::create_iterator(
+    const typename tsl::sparse_map<DataType, std::vector<RowID>>::const_iterator& it) {
+  return FlatMapIterator(std::make_unique<FlatMapIteratorImpl<DataType>>(it));
+}
+
```

Comment:
The thread is here: https://github.com/hyrise/hyrise/pull/2448#discussion_r1092009126

Reply:
gnar! Clicked a gazillion times GitHub's `Load more ...` to realize that you were faster.

Reply:
Hmm, not happy with it because it introduces a path where we need to explicitly name all supported data types. But I am fine with being outvoted here.

Reply:
I'm fine with all solutions.
Would you prefer the solution the `SegmentAccessor` uses?

Reply:
I would prefer a similar solution. But feel free to resolve. You have already discussed it and that looks like I am outvoted. :)